src/AppInstallerCommonCore/Public/winget/UserSettings.h

```diff
@@ -0,0 +1,66 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+#pragma once
+#include "AppInstallerStrings.h"
+#include "winget/settings/Source.h"
+#include "winget/settings/Visual.h"
+#include <json.h>
+
+#include <filesystem>
+#include <vector>
+#include <string>
+
+namespace AppInstaller::Settings
+{
+    // The type of argument.
+    enum class UserSettingsType
+    {
+        // Settings files don't exist. A file is created on the first call to the settings command.
+        Default,
+        // Loaded settings.json
+        Standard,
+        // Loaded settings.json.backup
+        Backup,
+    };
+
+    // Representation of the parsed settings file.
+    struct UserSettings
+    {
+        static UserSettings& Instance()
+        {
+            // If we go multi-threaded secure this.
+            static UserSettings userSettings;
+            return userSettings;
+        }
+
+        UserSettings(const UserSettings&) = delete;
+        UserSettings& operator=(const UserSettings&) = delete;
+
+        UserSettingsType GetType() const { return m_type; }
+        std::vector<std::string> GetWarnings() const { return m_warnings; }
+
+        void Reload();
+        void CreateFileIfNeeded();
+        void CreateBackup();
+
+        std::filesystem::path SettingsFilePath();
+        std::filesystem::path SettingsBackupFilePath();
+
+        // Settings
+        inline const Source& GetSource() const { return *m_source; }
+        inline const Visual& GetVisual() const { return *m_visual; }
```

Comment:
I'm not a fan of having to create classes and functions for every setting.  I would prefer a data driven model, thus requiring probably two files to be touched.  Will look at the code for parsing and see if that still seems like a reasonable position.

Reply:
So you kind of already do this, just not to the level that would make it easier to add new things.

There are many ways that it could be done, but for type safety I think it would need to be similar to the Execution::Context.

enum for settings
templated struct with specializations per enum
templated get and parse functions

To add a new setting:

add new enum value
add new specialization
add parse call in main Parse method

Templated struct contains:

json "path"
json data type
internal data type
default value
parse function to turn json type into internal type

Parsed data gets stored in a map similar to Execution::Context.  Get<enum>() looks there first, or returns struct<enum>::default if not present.

---
In reply to: [442441037](https://github.com/Microsoft/pkg/pull/436#discussion_r442441037) [](ancestors = 442441037)